### ISSUE TITLE
Possible performance degradation in Blaze client connection pool

### ISSUE THREAD

**Issue by afiore** [http4s/http4s#2340]
Hi there,

This is a follow-up to a help request [started on Gitter](https://gitter.im/http4s/http4s?at=5c2e201dbabbc178b21f985d). While upgrading a fairly large code-base (~30K lines) from http4s _0.15_ to _0.18_, we seem to have stumbled upon a performance degradation due to changes in the implementation of Blaze client connection pool, with some of our endpoints taking 3/4 times as long to return a response. We have put together two sample apps that reproduce the problem:

- [0.15.2a](http://github.com/afiore/http4s015connectionpool)
- [0.18.21](http://github.com/afiore/http4s018connectionpool)

The two apps perform many HTTP client calls in parallel, using a very lax and permissive connection pool configuration (assume this is an internal app, and has only one or two concurrent users). We observed that  the `0.15.2a` app performs `1000` concurrent `GET` requests to `https://example.com` in about _~2.07_ seconds, while the `0.18.21` app performs the same amount of calls in _~7.80_ seconds (in both cases, we took the median of 5 consecutive requests).

The endpoint can be run with the following curl command:

`curl -v localhost:8080/benchmark/client/run/1000`

We are not sure whether this significant difference in performance is due to us misconfiguring the client pool. In the `0.15.2a` app, the client is configured as follows:

```scala
    val client = PooledHttp1Client(
      maxTotalConnections = 50,
      BlazeClientConfig.defaultConfig.copy(
        idleTimeout = 59.second,
        requestTimeout = 60.seconds
      )
    )
```

while in the `0.18.21`:

```scala
    BlazeClientConfig.defaultConfig.copy(
      maxTotalConnections = 50,
      maxConnectionsPerRequestKey = _ => 50,
      requestTimeout = 60.seconds,
      idleTimeout = 59.seconds,
      responseHeaderTimeout = 599.seconds,
      maxWaitQueueLimit = 20000
    )
```
As the new `BlazeClientConfig` includes a few more options, we don't exclude the possibility of having significantly misconfigured it. However, several attempts at tweaking the parameters have held no significant improvement.

Thanks in advance,
Any help would be greatly appreciated!

**Comment by balatbn** [comment]
I ran both the tests in local box(Linux with JDK 1.8). During initial run, both clients took around 2-2.5 seconds. After that, 0.15 client takes around 300-400 ms. Meanwhile 0.18 client takes around 400-500 ms.

It looks like 0.18 is little bit slower, but I'm not able to reproduce the massive difference mentioned here.

What is the environment in which you are running the tests?

**Comment by rossabaker** [comment]
I'm on a Chromebook right now and can't run a good test, but would it be feasible to run the benchmark server on a remote app?  Or provide a different thread pool for client and server?

How many cores do these test machines have?

**Comment by cmcmteixeira** [comment]
I believe I may have found the issue.
As far as we were able to determine, http4s will often always close a connection meaning that each request will result in the allocation of a new connection.

(notice the difference in how stageShutdown is called in 0.18 and 0.15)
In version [0.18](https://github.com/http4s/http4s/blob/v0.18.22/blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala#L263) 
```scala
attributes -> rawBody.onFinalize(
  Stream
    .eval_(Async.shift(executionContext) *> F.delay {
      trailerCleanup(); cleanup(); stageShutdown()
    })
    .compile
    .drain
)
```
In version [0.15](https://github.com/http4s/http4s/blob/v0.15.16/blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala#L226):
```scala
attributes -> rawBody.onHalt {
  case End => Process.eval_(Task{ trailerCleanup(); cleanup(); }(executor))

  case c => Process.await(Task {
    logger.debug(c.asThrowable)("Response body halted. Closing connection.")
    stageShutdown()
  }(executor))(_ => Halt(c))
}
```

I haven't yet been able to determine why the `if(parser.contentComplete())` that surrounds this piece of code will for sometimes return true (concurrency is hard!) as internally it seems to use 
`_endOfContent == EndOfContent.END` which is only set by the `parser.reset()` call which  seems to be only called in the context of a Server Connection.

In terms of fixing this, reproducing the 0.15 behaviour should be enough; unless there's something that I'm not aware!

Let me know what you guys think :)

**Comment by cmcmteixeira** [comment]
I tried changing the code in question to:
```scala
attributes -> rawBody
              .onFinalize(
                Stream
                  .eval_(Async.shift(executionContext) *> F.delay {
                    trailerCleanup()
                    cleanup()
                  })
                  .compile
                  .drain
              )
              .handleErrorWith { _ =>
                Stream.eval_(Async.shift(executionContext) *> F.delay {
                  stageShutdown()
                })
              }
```
And we did have a major performance improvement from ~30s down to ~8s.
Using lsof we were "able" to more or less determine that the connections are now being reused rather than opened/closed every time.

**Comment by rossabaker** [comment]
Wow, yes, that's a great find.  I think the `Async.shift *> F.delay(...)` would be sufficient in the `onFinalize` block.  You could avoid the `.handleErrorWith` by using `.onFinalizeCase`, but I'm not sure it would be any cleaner.

**Comment by cmcmteixeira** [comment]
Cool! I'll raise the PR then
Should I use a specific version as base or just raise one for each version (0.18, 0.19,0.20) ?

**Comment by rossabaker** [comment]
Please target release-0.18.x.  0.19 is unmaintained (it was an errant tag).  And then we'll take care of merging to master, which will get it released on 0.20.

**Comment by cmcmteixeira** [comment]
This ticket can be closed now. It was fixed as part of the 0.18.23 release!

**Comment by rossabaker** [comment]
And, with tweaks, 0.20.0-M7. Thanks again!
